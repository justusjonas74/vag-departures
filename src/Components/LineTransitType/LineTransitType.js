import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './LineTransitType.css';

/* ----------------- */
/*    EXPORTS        */
/* ----------------- */

class LineTransitType extends Component {
  styleTransitType(departureTransitType) {
    const switchTerm = departureTransitType ? departureTransitType.toLowerCase() : ""
    switch (switchTerm) {
      case 'ubahn':
        return {
          cssClass: 'ubahn',
          symbolStr: 'fa-train'
        };
      case 'tram':
        return {
          cssClass: 'tram',
          symbolStr: 'fa-subway'
        };
      case 'bus':
        return {
          cssClass: 'bus',
          symbolStr: 'fa-bus'
        };
      default:
        return {
          cssClass: 'bus',
          symbolStr: 'fa-bus'
        };
    }
  }

  render() {
    const departureLine = this.props.departureLine
    const departureTransitType = this.props.departureTransitType;
    const typeStyles = this.styleTransitType(departureTransitType);

    return (
        <span className="departureTransitType">
          <i className={'fa ' + typeStyles.symbolStr} aria-hidden="true"></i>
          <span className={'departureLine '+ typeStyles.cssClass}>{departureLine}</span>
        </span>
    )
  }
}

LineTransitType.propTypes = {
  departureTransitType: PropTypes.string,
  departureLine: PropTypes.string
}

export default LineTransitType