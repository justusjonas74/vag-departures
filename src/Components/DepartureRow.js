import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './DepartureRow.css';

import LineTransitType from './LineTransitType';

/* ----------------- */
/*    EXPORTS        */
/* ----------------- */

class DepartureRow extends Component {
  render() {
    var cssDelayClass = 'isInTime'
    if (this.props.delay) {
      cssDelayClass = (this.props.delay.isDelayed ? 'isDelayed' : 'isInTime')
    }
    return (
      <tr className="departureRow">
          <td>
            <span className="departureTime">{this.props.departureTime}</span>
            <span className={"departureDelay " + cssDelayClass} >+{this.props.delay ? this.props.delay.time : "0"} </span></td>
          <LineTransitType
            departureLine={this.props.departureLine}
            departureTransitType={this.props.departureTransitType}
          />
          <td><span className="departureDirection">{this.props.departureDirection}</span></td>
      </tr>
    )
  }
}

DepartureRow.propTypes = {
  departureTime: PropTypes.string,
  delay:  PropTypes.shape({
    time: PropTypes.num,
    isDelayed: PropTypes.boolean
  }),
  departureLine: PropTypes.string,
  departureTransitType: PropTypes.string,
  departureDirection: PropTypes.string
}

export default DepartureRow