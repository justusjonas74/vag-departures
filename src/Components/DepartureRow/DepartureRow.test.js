import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import DepartureRow from './DepartureRow';
import LineTransitType from '../LineTransitType/LineTransitType';

describe('DepartureRow Component', () => {
  
  // BOILERPLATE
  let props;
  let mountedComponent;
  const departureRow = () => {
    if (!mountedComponent) {
      mountedComponent = mount(
        <DepartureRow {...props} />
      );
    }
    return mountedComponent;
  }

  beforeEach(() => {
    props = {
      departureTime: null,
      delay:  null,
      departureLine: null,
      departureTransitType: null,
      departureDirection: null
    };
    mountedComponent = undefined;
  });
  
  // TESTS
  it('should render correctly', () => {
    shallow(<DepartureRow />);
  });
  it("always renders a table cells", () => {
    const element = departureRow().find("td");
    expect(element.length).toBeGreaterThanOrEqual(1);
  });
  describe("the rendered table row", () => {
    // it("contains everything else that gets rendered", () => {
    //   const elements = departureRow().find("tr");
    //   const wrapper = elements.first();
    //   expect(wrapper).toEqual(departureRow().children());
    // });
    it('always includes one child of LineTransitType', ()=>{
      const elements = departureRow().find(LineTransitType);
      expect(elements.length).toEqual(1)
    })
    //SNAPSHOT
    it('renders correctly', () => {
       const tree = renderer
        .create(<DepartureRow {...props} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('should handle delayed trips correctly', ()=>{
    beforeEach(()=>{
      props.delay= {
        isDelayed:true,
        time: 12
      }
    })
    it('should apply special format', ()=>{
      const element = departureRow().find('.departureDelay')
      expect(element.hasClass('isDelayed')).toEqual(true)
    })
    it('should display the time difference correctly', ()=>{
      const element = departureRow().find('.departureDelay')
      expect(element.text()).toEqual('+12')
    })
  })
  describe('should handle non delayed trips correctly', ()=>{
    beforeEach(()=>{
      props.delay= {
        isDelayed:false,
        time: 0
      }
    })
    it('should apply special format', ()=>{
      const element = departureRow().find('.departureDelay')
      expect(element.hasClass('isInTime')).toEqual(true)
    })
    it('should display the time difference correctly', ()=>{
      const element = departureRow().find('.departureDelay')
      expect(element.text()).toEqual('+0')
    })
  })
  describe('should render correctly with all props', ()=>{
    beforeEach(()=>{
      props.departureTime = "12:12"
      props.departureLine = "U1"
      props.departureTransitType = "UBahn"
      props.departureDirection = "Langwasser"
      props.delay= {
        isDelayed:false,
        time: 0
      }
    })
    
    it('should display correct time', ()=>{
      const element = departureRow()
      const time = element.find('.departureTime').text()
      expect(time).toEqual('12:12')
    })
    it('should display correct direction', ()=>{
      const element = departureRow()
      const direction = element.find('.departureDirection').text()
      expect(direction).toEqual('Langwasser')
    })
  })
})