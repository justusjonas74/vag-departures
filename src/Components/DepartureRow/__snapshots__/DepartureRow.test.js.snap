// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepartureRow Component the rendered table row renders correctly 1`] = `
Array [
  <td>
    <span
      className="departureTime"
    />
    <span
      className="departureDelay isInTime"
    >
      +
      0
    </span>
  </td>,
  <td>
    <span
      className="departureTransitType"
    >
      <i
        aria-hidden="true"
        className="fa fa-bus"
      />
      <span
        className="departureLine bus"
      />
    </span>
  </td>,
  <td>
    <span
      className="departureDirection"
    />
  </td>,
]
`;
